presto-main/src/main/java/com/facebook/presto/operator/scalar/KleinbergBurstDetectionFunction.java

```diff
@@ -0,0 +1,219 @@
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.util.Failures.checkCondition;
+import static com.google.common.base.Preconditions.checkArgument;
+
+public final class KleinbergBurstDetectionFunction
+{
+    private KleinbergBurstDetectionFunction() {}
+
+    /**
+     * Base of the exponential distribution that is used for modeling the
+     * event frequencies. Must be greater than 1.0, and determines the granularity
+     * of burst intensities detected (greater value, lower and more compact
+     * intensities).
+     */
+    private static final double DEFAULT_KLEINBERG_S = 2.0;
+
+    /**
+     * Coefficient for the transition costs between states. Should be greater or
+     * equal than 1.0, although any value is accepted. High values make increases
+     * in the cost of intervals higher, amplifying big distances.
+     */
+    private static final double DEFAULT_KLEINBERG_GAMMA = 1.0;
+
+    @ScalarFunction
+    @Description("Return the Kleinberg burst detection for given timeSeries")
+    @SqlType("array(int)")
+    public static Block kleinbergBurstDetection(@SqlType("array(double)") Block timeSeries)
+    {
+        return kleinbergBurstDetection(timeSeries, DEFAULT_KLEINBERG_S, DEFAULT_KLEINBERG_GAMMA);
+    }
+
+    @ScalarFunction
+    @Description("Return the Kleinberg burst detection for given timeSeries, s and gamma")
+    @SqlType("array(int)")
+    public static Block kleinbergBurstDetection(
+            @SqlType("array(double)") Block timeSeries,
+            @SqlType(StandardTypes.DOUBLE) double s,
+            @SqlType(StandardTypes.DOUBLE) double gamma)
+    {
+        checkCondition(timeSeries.getPositionCount() > 1, INVALID_FUNCTION_ARGUMENT, "time series must have at least two values");
+        checkCondition(s > 1.0, INVALID_FUNCTION_ARGUMENT, "s: base of exponential distribution must be greater than 1");
+        checkCondition(gamma >= 1.0, INVALID_FUNCTION_ARGUMENT, "gamma: must be greater than or equal to 1");
+
+        return kleinberg(timeSeries, s, gamma);
+    }
+
+    /**
+     * Burst detection algorithm.
+     * @param timeSeries for which kleinberg burst detection need to be computed
+     * @param s base of the exponential distribution used for modeling the event frequencies
+     * @param gamma the coefficient for the transition costs between states
+     * @return burst intensity from each input timestamp in 'timeSeries'
+     */
+    private static Block kleinberg(Block timeSeries, double s, double gamma)
+    {
+        // The number of events [N].
+        int numEntries = timeSeries.getPositionCount();
+
+        // Calculate time intervals between successive events.
+        double[] intervals = new double[numEntries - 1];
+        // The minimum interval.
+        double delta = Double.MAX_VALUE;
+        for (int i = 0; i < numEntries - 1; i++) {
+            double timeCurrent = DOUBLE.getDouble(timeSeries, i);
```

Comment:
We need to check for null elements and either fail or return null. Also add a test for null elements.

Reply:
A null element in unit tests gets converted to 0. Is there any other way to test this?